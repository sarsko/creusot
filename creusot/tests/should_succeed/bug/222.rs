// SHOULD_PROVE Z3 NO_SPLIT
extern crate creusot_contracts;
use creusot_contracts::*;

trait A {
    #[logic]
    fn mktrue() -> Int {
        pearlite! { 5 }
    }

    #[law]
    #[ensures(Self::mktrue() <= 10)]
    fn is_true() {
        ()
    }
}

impl A for () {
    #[logic]
    fn mktrue() -> Int {
        pearlite! { 6 }
    }
}

// Should generate a VC for false
// impl A for bool {
//     #[logic]
//     fn mktrue() -> Int {
//         pearlite! { 15 }
//     }
// }
