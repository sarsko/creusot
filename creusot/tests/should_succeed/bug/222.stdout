module Type
  use Ref
  use mach.int.Int
  use prelude.Int8
  use prelude.Int16
  use mach.int.Int32
  use mach.int.Int64
  use prelude.UInt8
  use prelude.UInt16
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
end
module C222_A_Mktrue_Interface
  type self
  use mach.int.Int
  function mktrue (_ : ()) : int
end
module C222_A_Mktrue
  type self
  use mach.int.Int
  use mach.int.Int32
  function mktrue [#"../222.rs" 7 4 22] (_ : ()) : int =
    [#"../222.rs" 7 4 22] 5
end
module C222_A_IsTrue_Interface
  type self
  use mach.int.Int
  use mach.int.Int32
  clone C222_A_Mktrue_Interface as Mktrue0 with type self = self
  function is_true (_ : ()) : ()
end
module C222_A_IsTrue
  type self
  use mach.int.Int
  use mach.int.Int32
  clone C222_A_Mktrue_Interface as Mktrue0 with type self = self
  function is_true [#"../222.rs" 13 4 16] (_ : ()) : () =
    [#"../222.rs" 13 4 16] ()
  axiom is_true_spec : [#"../222.rs" 12 4 36] Mktrue0.mktrue () <= 10
end
module C222_A_IsTrue_Impl
  type self
  use mach.int.Int
  use mach.int.Int32
  clone C222_A_Mktrue as Mktrue0 with type self = self
  let rec ghost function is_true (_ : ()) : ()
    ensures { [#"../222.rs" 12 4 36] Mktrue0.mktrue () <= 10 }
    
   =
    [#"../222.rs" 13 4 16] ()
end
module C222_Impl0_Mktrue_Interface
  use mach.int.Int
  function mktrue (_ : ()) : int
end
module C222_Impl0_Mktrue
  use mach.int.Int
  use mach.int.Int32
  function mktrue [#"../222.rs" 20 4 22] (_ : ()) : int =
    [#"../222.rs" 20 4 22] 6
end
module C222_Impl0
  clone C222_Impl0_Mktrue as Mktrue0
  clone C222_A_IsTrue as IsTrue0 with type self = (), function Mktrue0.mktrue = Mktrue0.mktrue
  clone C222_A_Mktrue_Interface as Mktrue1 with type self = (), function mktrue = Mktrue0.mktrue
end
